### PR TITLE
Add hdrf design doc

### PR DESCRIPTION
**Pull request by ggaaooppeenngg**
Signed-off-by: Peng Gao <peng.gao.dut@gmail.com>

### CODE REVIEW

File: docs/design/hdrf.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+To represent a more complex hierarchical group sharing other than a big weighted flattened namespaces sharing, the HDRF(hierarchical dominant resource fairness) may need to be adopted.
+The original [paper](https://people.eecs.berkeley.edu/~alig/papers/h-drf.pdf) refers to two points that naive DRF does not suit in.
+The first is the starvation caused by children with a complementary dominant resource.
+```
+              n
+    /                   \
+  n1 w=50%           n2 w=50%
+(0 CPU,1 GPU)
+                  /            \
+		        n2,1  w=50%    n2,2 w=50%
```
Comment (hzxuzhonghu) [inline docs/design/hdrf.md:17]: please update indents

File: docs/design/hdrf.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+a 100% share to its parent n2 group. Consider a job removed, and a new job added, then sibling n2,2 group with different dominant share(for example GPU 50%) will be punished since GPU will be always allocated to the n1 group which has a smaller dominant share than the n2 group. To avoid this starvation, HDRF rescales children to the minimum node. In this case, the n2,1 group will be scaled to (10,0) * (0.5/1) = (5,0), it is then summed to the parent, thus the parent n2 group will have a share of 50% which is our desired state.
+
+The second is nodes blocking caused by a saturated node.
+```
+                                n
+/                    |                     |               \
+n1                   n2                    n3                n4
+(1 CPU,0 GPU)  (1 CPU,0 GPU)         /          \          (0 CPU,1 GPU)
+                                     n3,1          n3,2
+			                     (1 CPU,0 GPU)  (0 CPU,1 GPU)
```
Comment (hzxuzhonghu) [inline docs/design/hdrf.md:31]: The graph looks bad too

**Review by hzxuzhonghu — COMMENTED** [review]
Looks great

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by k82cn** [comment]
is there any limitation of this feature?

**Comment by ggaaooppeenngg** [comment]
@k82cn  
1. This feature conflicts with the proportion plugin. If the hdrf is enabled, the proportion should be disabled, and a reclaim function which compares queue by hierarchical shares and weights should be added. 
2. When a job is added in the tree, hdrf travels through the root recursively since it needs to mark the all saturated nodes, if some resources are fully allocated. This may be ineffective.

**Comment by ggaaooppeenngg** [comment]
@k82cn  
1. This feature conflicts with the proportion plugin. If the hdrf is enabled, the proportion should be disabled, and a reclaim function which compares queue by hierarchical shares and weights should be added. 
2. When a job is added in the tree, hdrf travels through the root recursively since it needs to mark all the saturated nodes, if some resources are fully allocated. This may be ineffective.

**Comment by k82cn** [comment]
> @k82cn
> 
> 1. This feature conflicts with the proportion plugin. If the hdrf is enabled, the proportion should be disabled, and a reclaim function which compares queue by hierarchical shares and weights should be added.
> 2. When a job is added in the tree, hdrf travels through the root recursively since it needs to mark all the saturated nodes, if some resources are fully allocated. This may be ineffective.

hm... got that, can you help to add those limitation into the doc?

@hzxuzhonghu @Thor-wl , please help to work out a plan to close those limitations.

/cc @william-wang @kevin-wangzefeng

**Comment by k82cn** [comment]
LGTM overall, let's consider this as an alpha feature :)

**Pull request merged**